
<!doctype html>
<html lang="en" class="no-js">
  <head>

      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">

        <meta name="description" content="This page documents everything that has been done/generated in this project">




      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.1.0">



        <title>Primary Analysis - TotalSeq Analysis Methods and Results For Healthy Control Vs Uveitis Patients</title>



      <link rel="stylesheet" href="assets/stylesheets/main.33e2939f.min.css">


        <link rel="stylesheet" href="assets/stylesheets/palette.ef6f36e2.min.css">



          <meta name="theme-color" content="#ef5552">







        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>




      <link rel="stylesheet" href="stylesheets/extra.css">





  </head>







    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="red" data-md-color-accent="">


    <script>function __prefix(e){return new URL(".",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>

    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">


        <a href="#a-prepare-sampleinformationtab-file" class="md-skip">
          Skip to content
        </a>

    </div>
    <div data-md-component="announce">

    </div>

      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="TotalSeq Analysis Methods and Results For Healthy Control Vs Uveitis Patients" class="md-header__button md-logo" aria-label="TotalSeq Analysis Methods and Results For Healthy Control Vs Uveitis Patients" data-md-component="logo">


  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TotalSeq Analysis Methods and Results For Healthy Control Vs Uveitis Patients
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">

              Primary Analysis

          </span>
        </div>
      </div>
    </div>



      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>

<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>


  </nav>
</header>

    <div class="md-container" data-md-component="container">




      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">



              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">





<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="TotalSeq Analysis Methods and Results For Healthy Control Vs Uveitis Patients" class="md-nav__button md-logo" aria-label="TotalSeq Analysis Methods and Results For Healthy Control Vs Uveitis Patients" data-md-component="logo">


  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    TotalSeq Analysis Methods and Results For Healthy Control Vs Uveitis Patients
  </label>

  <ul class="md-nav__list" data-md-scrollfix>








    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        Home
      </a>
    </li>












    <li class="md-nav__item md-nav__item--active">

      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">


        <label class="md-nav__link md-nav__link--active" for="__toc">
          Primary Analysis
          <span class="md-nav__icon md-icon"></span>
        </label>

      <a href="primary-analysis.html" class="md-nav__link md-nav__link--active">
        Primary Analysis
      </a>


<nav class="md-nav md-nav--secondary" aria-label="Table of contents">



    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>

        <li class="md-nav__item">
  <a href="#a-prepare-sampleinformationtab-file" class="md-nav__link">
    A. Prepare SampleInformation.tab file
  </a>

</li>

        <li class="md-nav__item">
  <a href="#b-prepare-the-input-data-files-output-folders-and-the-swarm-job-file" class="md-nav__link">
    B. Prepare the input data files, output folders and the swarm job file
  </a>

</li>

        <li class="md-nav__item">
  <a href="#c-primary-analysis-barcodeumicell-counting-using-cellranger" class="md-nav__link">
    C. Primary analysis - barcode/umi/cell counting using cellranger
  </a>

</li>

        <li class="md-nav__item">
  <a href="#d-organize-the-outputs-and-cleanup" class="md-nav__link">
    D. Organize the outputs and cleanup
  </a>

</li>

        <li class="md-nav__item">
  <a href="#e-aggregate-cellranger-summary-outputs" class="md-nav__link">
    E. Aggregate cellranger summary outputs
  </a>

</li>

        <li class="md-nav__item">
  <a href="#software" class="md-nav__link">
    Software
  </a>

</li>

        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    References
  </a>

</li>

    </ul>

</nav>

    </li>










    <li class="md-nav__item">
      <a href="qualityControl.html" class="md-nav__link">
        Quality Control
      </a>
    </li>










    <li class="md-nav__item">
      <a href="normalization.html" class="md-nav__link">
        Normalization
      </a>
    </li>










    <li class="md-nav__item">
      <a href="clustering.html" class="md-nav__link">
        Clustering
      </a>
    </li>










    <li class="md-nav__item">
      <a href="differential.html" class="md-nav__link">
        Marker Discovery
      </a>
    </li>










    <li class="md-nav__item">
      <a href="annotation.html" class="md-nav__link">
        Annotation
      </a>
    </li>



  </ul>
</nav>
                  </div>
                </div>
              </div>



          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">



                  <h1>Primary Analysis</h1>

                <h3 id="a-prepare-sampleinformationtab-file">A. Prepare SampleInformation.tab file</h3>
<ol>
<li>Current SampleInformation.tab file is a tab delimited text file. This file could be edited in a plain text editor or a spreadsheet program. Update this file when a new sample is added to the project.</li>
</ol>
<h3 id="b-prepare-the-input-data-files-output-folders-and-the-swarm-job-file">B. Prepare the input data files, output folders and the swarm job file</h3>
<ol>
<li>Make sure you have the Results/cellranger folder - to store your results</li>
<li>The Tools/prep.sh file is a bash script file that should be run first, to prepare files/folders and to generate the swarm job file.</li>
<li>Open the prep.sh file in a plain text editor and edit the "samples" variable to update new sample names as it is entered in the SampleInformation.tab file.</li>
<li>Open a shell/terminal, to change to the Tools directory, run the below command
<div class="highlight"><pre><span></span><code>cd /data/../TotalSeq/Tools/
</code></pre></div></li>
<li>Run the prep.sh script using the below command
<div class="highlight"><pre><span></span><code>bash prep.sh
</code></pre></div></li>
<li>Check the Results folder to make sure you have directories for new samples and that the directory has corresponding fastq files and library file</li>
<li>Check the Tools folder to make sure you have the "totalseqall.swarm" file</li>
</ol>
<h3 id="c-primary-analysis-barcodeumicell-counting-using-cellranger">C. Primary analysis - barcode/umi/cell counting using cellranger</h3>
<ol>
<li>Change to the Tools directory using the below command;
<div class="highlight"><pre><span></span><code>cd /data/../TotalSeq/Tools/
</code></pre></div></li>
<li>Run the swarm job using the below command;
<div class="highlight"><pre><span></span><code>swarm -f totalseqall.swarm -g 64 -t 28 --gres=lscratch:200 --partition=norm --module cellranger
</code></pre></div></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note down your job id</p>
</div>
<ol>
<li>Check your job status using the "jobload" command or the squeue command;
<div class="highlight"><pre><span></span><code>squeue --job 14354599
</code></pre></div></li>
<li>Extend the swarm job running time using the below command;
<div class="highlight"><pre><span></span><code>newwall --jobid 13558558 --time 10:00:00
</code></pre></div></li>
</ol>
<h3 id="d-organize-the-outputs-and-cleanup">D. Organize the outputs and cleanup</h3>
<ol>
<li>Once the swarm job is complete (should take about three hours), change to the Tools directory using the below command;
<div class="highlight"><pre><span></span><code>cd /data/../TotalSeq/Tools/
</code></pre></div></li>
<li>Open the "organizeout.sh" script in text editor and updated the variable "samples" with your new sample names</li>
<li>Run the "organizeout.sh" script using the below command;
<div class="highlight"><pre><span></span><code>bash organizeout.sh
</code></pre></div></li>
<li>Check out the Results folder to make sure you have the expected output files.</li>
</ol>
<div class="admonition example">
<p class="admonition-title">Example CellRanger Output location for Sample NS3R189BTS</p>
<p><a href="https://github.com/PulakNath/bcr-uveitis/tree/main/Results/cellranger/NS3R189BTS/cellranger_output/">/data/../TotalSeq/Results/cellranger/NS3R189BTS/cellranger_output/</a></p>
</div>
<h3 id="e-aggregate-cellranger-summary-outputs">E. Aggregate cellranger summary outputs</h3>
<ol>
<li>Login to biowulf and request an interactive node using the below 'sinteractive' command, with 64 cores, 160gb RAM and 300gb of storage;
<div class="highlight"><pre><span></span><code>sinteractive --cpus-per-task=64 --mem=160g --gres=lscratch:300
</code></pre></div></li>
<li>Once the interacive node is assigned, change to the Tools directory using the below command;
<div class="highlight"><pre><span></span><code>cd /data/../TotalSeq/Tools/
</code></pre></div></li>
<li>Load the R module using the below command;
<div class="highlight"><pre><span></span><code>module load R
</code></pre></div></li>
<li>Once the R module is loaded, run the 'cellrangerReport.R' script using the below command;
<div class="highlight"><pre><span></span><code>Rscript cellrangerReport.R
</code></pre></div></li>
<li>The cellranger report is now available in the "Reports" folder. A link to this report is also given <a href="cellrangerReport.html">here</a></li>
</ol>
<h3 id="software">Software</h3>
<ol>
<li>cellranger  6.0.0</li>
<li>R 4.0.5</li>
<li>mkdocs 1.1.2</li>
</ol>
<h3 id="references">References</h3>
<ol>
<li><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/feature-bc-analysis">Feature reference file generation</a></li>
<li><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ct">TotalSeq</a></li>
<li><a href="https://hpc.nih.gov/apps/cellranger.html">CellRanger Biowulf</a></li>
<li><a href="https://hpc.nih.gov/apps/swarm.html">Swarm Biowulf</a></li>
</ol>







            </article>
          </div>
        </div>

      </main>


<footer class="md-footer">

    <nav class="md-footer__inner md-grid" aria-label="Footer">

        <a href="index.html" class="md-footer__link md-footer__link--prev" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Home
            </div>
          </div>
        </a>


        <a href="qualityControl.html" class="md-footer__link md-footer__link--next" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Quality Control
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>

    </nav>

  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">

        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>

      </div>

    </div>
  </div>
</footer>

    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": ".", "features": "toc.integrate", "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "assets/javascripts/workers/search.fe42c31b.min.js", "version": null}</script>


      <script src="assets/javascripts/bundle.d892486b.min.js"></script>


  </body>
</html>
